src/avcodec/bitstream.rs

```diff
@@ -0,0 +1,143 @@
+            match ffi::av_bsf_init(self.as_mut_ptr()).upgrade() {
+                Ok(_) => Ok(()),
+                Err(x) => Err(RsmpegError::BitstreamInitializationError(x)),
+            }
+        }
+    }
+    /// Provide input data for the bitstream filter to process. To signal the end of the stream, send an NULL packet to the filter.
+    ///
+    /// See [`ffi::av_bsf_send_packet`] for more info.
+    pub fn send_packet(&mut self, packet: Option<ffi::AVPacket>) -> Result<()> {
+        // TODO: Ensure init is called first
+        let packet_ptr = match packet {
+            Some(mut packet) => &mut packet,
+            None => std::ptr::null_mut(),
+        };
+
+        match unsafe { ffi::av_bsf_send_packet(self.as_mut_ptr(), packet_ptr) }.upgrade() {
+            Ok(_) => Ok(()),
+            Err(AVERROR_EAGAIN) => Err(RsmpegError::BitstreamSendPacketAgainError),
+            Err(ffi::AVERROR_EOF) => Err(RsmpegError::BitstreamFlushedError),
+            Err(x) => Err(RsmpegError::BitstreamSendPacketError(x)),
+        }
+    }
+    /// Get processed data from the bitstream filter.
+    ///
+    /// See [`ffi::av_bsf_receive_packet`] for more info.
+    pub fn receive_packet(&mut self, mut packet: ffi::AVPacket) -> Result<ffi::AVPacket> {
+        match unsafe { ffi::av_bsf_receive_packet(self.as_mut_ptr(), &mut packet) }.upgrade() {
+            Ok(_) => Ok(packet),
+            Err(AVERROR_EAGAIN) => Err(RsmpegError::BitstreamSendPacketAgainError),
+            Err(ffi::AVERROR_EOF) => Err(RsmpegError::BitstreamFlushedError),
+            Err(x) => Err(RsmpegError::BitstreamReceivePacketError(x)),
+        }
+    }
+    /// See [`ffi::av_bsf_flush`] for more info.
+    pub fn flush(&mut self) {
+        unsafe {
+            ffi::av_bsf_flush(self.as_mut_ptr());
+        }
+    }
+    /// Copies `source_params` into [`ffi::AVBSFContext`]'s `par_in` field. So we only need a reference to `source_params`.
+    ///
+    /// See [`ffi::avcodec_parameters_copy`] for more info.
+    pub fn set_par_in(&mut self, source_params: &ffi::AVCodecParameters) -> Result<()> {
+        match unsafe { ffi::avcodec_parameters_copy(self.par_in, source_params) }.upgrade() {
+            Ok(_) => Ok(()),
+            Err(e) => Err(RsmpegError::AVError(e)),
+        }
+    }
+    // FIXME: Returns bsf_list filter for some reason...
+    #[doc(hidden)]
+    pub fn get_null() -> Self {
+        let mut bsfc_raw = ptr::null_mut();
+
+        unsafe {
+            ffi::av_bsf_get_null_filter(&mut bsfc_raw);
+            AVBSFContext::from_raw(bsfc_raw.upgrade().unwrap())
+        }
+    }
+    // fn is_init(&self) -> bool {
+    // TODO: This will be used in receive/send_packet
+    // See https://github.com/FFmpeg/FFmpeg/blob/bc70684e74a185d7b80c8b80bdedda659cb581b8/libavcodec/bsf.c#L145
+    // }
+}
+
+impl Drop for AVBSFContext {
+    fn drop(&mut self) {
+        unsafe { ffi::av_bsf_free(&mut self.as_mut_ptr()) }
+    }
+}
+
```

Comment:
Please add a `#[cfg(test)]` here. https://doc.rust-lang.org/book/ch11-03-test-organization.html#the-tests-module-and-cfgtest